packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx

```diff
@@ -40,11 +43,22 @@ const imageNodeStyle = (
   return imageNodeStyle;
 };
 
-export const ImageNode = memo(({ data, isConnectable, id, selected }: NodeProps<ImageNodeData>) => {
+export const ImageNode = memo(({ data, id, selected }: NodeProps<ImageNodeData>) => {
   const theme = useTheme();
   const { dropFeedbackStyleProvider } = useDropNode();
   const { httpOrigin } = useContext<ServerContextValue>(ServerContext);
-  const { onConnectionStartElementClick, newConnectionStyleProvider } = useConnector();
+  const { newConnectionStyleProvider } = useConnector();
+  const updateNodeInternals = useUpdateNodeInternals();
+
+  const firstUpdate = useRef<boolean>(true);
+
+  useEffect(() => {
+    if (firstUpdate.current && firstUpdate) {
+      firstUpdate.current = false;
+    } else {
+      updateNodeInternals(id);
+    }
+  }, [JSON.stringify(data.connectionHandles)]);
```

Comment:
- This code should not be there, we don't want to repeat that code in every single node, this belongs in a dedicated handle hook. Custom shapes will have to reuse it for example.
- Why do you need this ref? Why not a state?

I agree with @pcdavid, on top of being a potential performance issue, you do not have any idea of what would this behavior be in the future. That being said, it works for a large number of use cases it can be a pretty code solutions BUT, and I say this "but" with every single letter in upper case, I think that the unique place where this code would be needs a link to the original explanation which details why we need to be VERY VERY careful with this kind of solutions: https://github.com/facebook/react/issues/14476#issuecomment-471199055

Honestly, given the use case, I would rather have something like this:


```
const connectionHandlesIdentity = data.connectionHandles
                                      .map(handle => `${handle.nodeId}#${handle.edgeId}`)
                                      .join(', ');

useEffect(() => { ... }, [connectionHandlesIdentity])
```

The main difference being that if we change the structure of ConnectionHandle, we can select whether or not this code should be impacted.

Reply:
Thanks for the explanation, I'll use a string as dependency of the useffect and put it in a custom hook then.

I used a ref because a state change would trigger a re render (unlike a ref change) and call updateNodeInternals which if called too much can impact the performance a lot. From my testing in this particular use case I get more render by using a state than a ref.